### ISSUE TITLE
VI fit surrogate "stateless sampling" becomes slow using while_loops

### ISSUE THREAD

**Issue by jguhlin** [tensorflow/probability#1448]
I'm using the nightly (but tested on most recent releases) and getting this error with no obvious way to fix:

```
Saw Tensor seed Tensor("seed:0", shape=(2,), dtype=int32), implying stateless sampling. Autovectorized functions that use 
stateless sampling may be quite slow because the current implementation falls back to an explicit loop. This will be fixed in the 
future. For now, you will likely see better performance from stateful sampling, which you can invoke by passing a Python `int` seed.
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
WARNING:tensorflow:Using a while_loop for converting StatelessRandomUniformFullIntV2
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
WARNING:tensorflow:Using a while_loop for converting StatelessRandomNormalV2
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
WARNING:tensorflow:Using a while_loop for converting StatelessRandomUniformFullIntV2
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
WARNING:tensorflow:Using a while_loop for converting StatelessRandomNormalV2
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
WARNING:tensorflow:Using a while_loop for converting StatelessRandomUniformFullIntV2
WARNING:tensorflow:Using a while_loop for converting StatelessRandomGetKeyCounter
.......
```

My code looks like:
```
losses = tfp.vi.fit_surrogate_posterior(
    formula.unnormalized_log_prob,
    surrogate,
    sample_size=122,
    optimizer=tf.optimizers.Adam(learning_rate=1e-2),
    num_steps=8192*4,
    seed=42
)
```

So even passing a python int does not change this. It's also very slow (as the warning stated) compared to prior versions. I've tried switching out to JAX but the vi modules are not in the JAX substrate API. Are they still supported or not yet?

Cheers

**Comment by davmre** [comment]
TF's lack of support for vectorizing stateless samplers is unfortunately a pretty big pain point at the moment. You're right that there's no straightforward fix, though there are a few possible workarounds.

To be precise, the slowdown occurs when drawing multiple samples by applying `vectorized_map` to a TF stateless sampling op. So the workarounds essentially cluster into 
1. don't use `vectorized_map`, or 
2. don't draw multiple samples, or 
3. don't use stateless sampling, or
4. don't use TF.

The workaround I'd generally recommend (because it's reliable and involves the fewest compromises) is to avoid `vectorized_map`, by converting your surrogate posterior to either a non-autobatched joint distribution or an autobatched JD that sets `use_vectorized_map=False`. The cost is that your model needs to be written flexibly enough to work with potential batch dimensions that arise from drawing multiple samples (vs under `vectorized_map` where the batch dimension is invisible to the model). If it's not clear how to do this, feel free to post how you're building the surrogate and we can try to advise.

For completeness, here are some other potential workarounds:

Specifically to VI, you can avoid drawing multiple samples by passing `sample_size=1`, which will bypass the issue, at the cost of making the individual steps noisier (but also cheaper). 

Alternatively, you can attempt to force TFP to use stateful sampling ops, by passing `int`- rather than Tensor-valued seeds. I don't really recommend this, because some code paths will silently convert these into stateless seeds, and there are no guarantees that any given code path *won't* start doing this conversion in the future. For example, `fit_surrogate_posterior`  now converts all seeds to stateless so that they can be passed through the optimization loop (implemented in `tfp.math.minimize`), which is why passing an `int`-valued seed isn't making a difference for you. For now, you could likely bypass this  by writing your own optimization loop that invokes `tfp.vi.monte_carlo_variational_loss` directly, but this is not a future-proof approach.

Finally, you could consider switching to the JAX backend if it suits your needs, since all JAX sampling ops are stateless, and JAX's `vmap` supports vectorizing them. Unfortunately, the VI APIs in TFP are designed around the concept of trainable variables, which has no direct JAX equivalent, so porting them creates some new design problems. We'd like to come up with a good design (and help/suggestions always appreciated!), but there's currently not a definite plan or timeframe, so for now I'd encourage the JAX solution only if you're comfortable writing your own variational losses and optimization loops.
